twoliter/embedded/img2img

```diff
@@ -1,3 +1,318 @@
+  *)
+    echo "unexpected arg: ${opt}" >&2
+    exit 1
+    ;;
+  esac
+done
+
+WORKDIR="$(mktemp -d)"
+pushd "${WORKDIR}" >/dev/null
+
+# Clean up working directory to reduce size of layer.
+cleanup() {
+  [[ -n "${WORKDIR}" ]] && rm -rf "${WORKDIR}"
+}
+trap 'cleanup' EXIT
+
+# Import the partition helper functions.
+# shellcheck source=partyplanner
+. "${0%/*}/partyplanner"
+
+# Import the image helper functions.
+# shellcheck source=imghelper
+. "${0%/*}/imghelper"
+
+# Validate that the values for the args are sane.
+sanity_checks \
+  "${OUTPUT_FMT}" "${PARTITION_PLAN}" "${OVF_TEMPLATE}" "${UEFI_SECURE_BOOT}"
+
+###############################################################################
+# Section 1: prepare working environment
+
+# Store output artifacts in a versioned directory.
+OUTPUT_DIR="${OUTPUT_DIR}/${VERSION_ID}-${BUILD_ID}"
+mkdir -p "${OUTPUT_DIR}"
+
+declare OS_IMAGE DATA_IMAGE
+BOOT_IMAGE="${WORKDIR}/boot.ext4"
+VERITY_IMAGE="${WORKDIR}/root.verity"
+ROOT_IMAGE="${WORKDIR}/root.ext4"
+EFI_IMAGE="${WORKDIR}/efi.vfat"
+
+ROOT_MOUNT="$(mktemp -p "${WORKDIR}" -d root.XXXXXXXXXX)"
+BOOT_MOUNT="$(mktemp -p "${WORKDIR}" -d boot.XXXXXXXXXX)"
+EFI_MOUNT="$(mktemp -p "${WORKDIR}" -d efi.XXXXXXXXXX)"
+
+declare -A partsize partoff
+set_partition_sizes \
+  "${OS_IMAGE_SIZE_GIB}" "${DATA_IMAGE_SIZE_GIB}" "${PARTITION_PLAN}" \
+  partsize partoff
+
+for part in EFI-A BOOT-A ROOT-A HASH-A; do
+  # Each partition is aligned to a 1 MiB boundary, and extends to the sector
+  # before the next partition starts. Specify the end point in sectors so we
+  # can subtract a sector to fix the off-by-one error that comes from adding
+  # start and size together. (1 MiB contains 2048 512-byte sectors.)
+  part_start="${partoff["${part}"]}"
+  part_end="$((part_start + partsize["${part}"]))"
+  part_end="$((part_end * 2048 - 1))"
+done
+
+# Process and stage the input images to the working directory.
+stage_images "${INPUT_DIR}" "${OUTPUT_FMT}" OS_IMAGE DATA_IMAGE
+
+# Validate that our input image matches the partition plan.
+SGDISK_OUTPUT="$(sgdisk --print "${OS_IMAGE}")"
+for part in EFI-A BOOT-A ROOT-A HASH-A; do
+  case "${part}" in
+  EFI-A)
+    sgdisk_start="$(awk /EFI-SYSTEM/'{print $2}' <<<"${SGDISK_OUTPUT}")"
+    sgdisk_size="$(awk /EFI-SYSTEM/'{print $4}' <<<"${SGDISK_OUTPUT}")"
+    ;;
+  *)
+    sgdisk_start="$(awk /BOTTLEROCKET-"${part}"/'{print $2}' \
+      <<<"${SGDISK_OUTPUT}")"
+    sgdisk_size="$(awk /BOTTLEROCKET-"${part}"/'{print $4}' \
+      <<<"${SGDISK_OUTPUT}")"
+    ;;
+  esac
+
+  # 1 MiB contains 2048 512-byte sectors.
+  sgdisk_start_mib="$((sgdisk_start / 2048))"
+
+  if [[ "${partoff["${part}"]}" -ne "${sgdisk_start_mib}" ]]; then
+    echo "start mismatch between partition plan and disk: '${part}'" >&2
+    exit 1
+  fi
+
+  if [[ "${partsize["${part}"]}" -ne "${sgdisk_size%.0}" ]]; then
+    echo "size mismatch between partition plan and disk: '${part}'" >&2
+    exit 1
+  fi
+done
+
```

Comment:
I'm probably going to replace this block with a new `get_partition_sizes` function in `partyplanner` and compare one array to the other.

Reply:
Replaced.